### PR TITLE
expression, util: 1. refine castStrAsReal 2. cast use new architecture

### PR DESCRIPTION
**Pull request by XuHuaiyu**
This commit does 2 things:
1. use types.StrToFloat instead of strconv.ParseFloat when cast str as real
2. use new architecture to evaluate cast

PTAL @shenli @hanfei1991 @zz-jason @winkyao

### CODE REVIEW

File: expression/builtin_cast.go [hunk 1]
```diff
@@ -659,14 +659,20 @@ type builtinCastStringAsRealSig struct {
...
 	}
 	val, isNull, err := b.args[0].EvalString(row, b.getCtx().GetSessionVars().StmtCtx)
 	if isNull || err != nil {
 		return res, isNull, errors.Trace(err)
 	}
-	res, err = strconv.ParseFloat(val, 64)
+	res, err = types.StrToFloat(sc, val)
+	var err1 error
+	res, err1 = types.ProduceFloatWithSpecifiedTp(res, b.tp, sc)
+	if err == nil && err1 != nil {
```
Comment (hanfei1991) [inline expression/builtin_cast.go]: no need to code like this

File: expression/builtin_cast.go [hunk 2]
```diff
@@ -659,14 +659,20 @@ type builtinCastStringAsRealSig struct {
 }
 
 func (b *builtinCastStringAsRealSig) evalReal(row []types.Datum) (res float64, isNull bool, err error) {
+	sc := b.getCtx().GetSessionVars().StmtCtx
 	if IsHybridType(b.args[0]) {
-		return b.args[0].EvalReal(row, b.getCtx().GetSessionVars().StmtCtx)
+		return b.args[0].EvalReal(row, sc)
 	}
 	val, isNull, err := b.args[0].EvalString(row, b.getCtx().GetSessionVars().StmtCtx)
```
Comment (lamxTyler) [inline expression/builtin_cast.go]: This can also be replaced with `sc`?

**Comment by XuHuaiyu** [comment]
PTAL 
@hanfei1991 @lamxTyler  @zz-jason

File: util/types/datum.go [hunk 3]
```diff
@@ -759,13 +759,26 @@ func (d *Datum) convertToFloat(sc *variable.StatementContext, target *FieldType)
 	default:
 		return invalidConv(d, target.Tp)
 	}
+	var err1 error
+	f, err1 = ProduceFloatWithSpecifiedTp(f, target, sc)
+	if err == nil && err1 != nil {
```
Comment (hanfei1991) [inline util/types/datum.go]: Still no need to code like that.

**Comment by hanfei1991** [comment]
LGTM

File: util/types/datum.go (hunk 3 above)
Comment (XuHuaiyu) [inline util/types/datum.go]: Some old [test cases](https://github.com/pingcap/tidb/blob/master/util/types/convert_test.go#L512) depend on this implementation,
so I keep it here.

**Comment by zz-jason** [comment]
LGTM

**Pull request merged**